### PR TITLE
Add (random) seed api

### LINKED ISSUE #14: Slow deserialization compared to stock `HashMap`
I tried using `FxHashMap` to improve performance of internal hash maps, but I noticed that it takes several times longer to deserialize a large hash map that we need to persist. I couldn't find anything about it on google, so I created a minimal example and am reporting it here.

The behavior can be reproduced using a simple set of random elements. For example, on my machine this code takes 0.08s to serialize and 1.2s to deserialize a `HashSet`, as reported by `cargo run --release`:

```rust
use std::collections::HashSet;
use std::time::Instant;
use rand::Rng;

fn main() {
    let mut rnd = rand::thread_rng();
    let h: HashSet<u64> = (0..10_000_000).map(|_| rnd.gen::<u64>()).collect();

    let t0 = Instant::now();
    let mut out = vec![];
    bincode::serialize_into(&mut out, &h).unwrap();
    let t1 = Instant::now();
    println!("serialize: {}", (t1 - t0).as_secs_f64());

    let h2: HashSet<u64> = bincode::deserialize_from(&*out).unwrap();
    let t2 = Instant::now();
    println!("deserialize: {}", (t2 - t1).as_secs_f64());

    println!("{}", h2.len());
}
```

Trivially changing `HashSet` to `rustc_hash::FxHashSet` increases deserialization time to 5.5s (almost 5x slower), while serialization is unchanged. In our actual use case the original deserialization takes on the order of 2 minutes, so the slowdown visibly affects our total processing times. Code:

```rust
use rustc_hash::FxHashSet;
use std::time::Instant;
use rand::Rng;

fn main() {
    let mut rnd = rand::thread_rng();
    let h: FxHashSet<u64> = (0..10_000_000).map(|_| rnd.gen::<u64>()).collect();

    let t0 = Instant::now();
    let mut out = vec![];
    bincode::serialize_into(&mut out, &h).unwrap();
    let t1 = Instant::now();
    println!("serialize: {}", (t1 - t0).as_secs_f64());

    let h2: FxHashSet<u64> = bincode::deserialize_from(&*out).unwrap();
    let t2 = Instant::now();
    println!("deserialize: {}", (t2 - t1).as_secs_f64());

    println!("{}", h2.len());
}
```

I thought that perhaps the issue is in pathological behavior when rebuilding a hashmap from elements extracted in the same order, so I modified the code to serialize a `Vec` and deserialize by building an `FxHashSet` out of the `Vec`. That results in slightly slower serialization of 0.12s, but deserializes in just 0.15s, which includes both the time to deserialize the vector and the time to collect it into a new `FxHashSet`. (Applying this to ordinary `HashSet` didn't speed it up, it takes 0.11s to serialize and 1.3s to deserialize.) Code:

```rust
use rustc_hash::FxHashSet;
use std::time::Instant;
use rand::Rng;

fn main() {
    let mut rnd = rand::thread_rng();
    let h: FxHashSet<u64> = (0..10_000_000).map(|_| rnd.gen::<u64>()).collect();

    let t0 = Instant::now();
    let mut out = vec![];
    let hack = h.iter().copied().collect::<Vec<u64>>();
    bincode::serialize_into(&mut out, &hack).unwrap();
    let t1 = Instant::now();
    println!("serialize: {}", (t1 - t0).as_secs_f64());

    let hack2: Vec<u64> = bincode::deserialize_from(&*out).unwrap();
    let h2: FxHashSet<u64> = hack2.into_iter().collect();
    let t2 = Instant::now();
    println!("deserialize: {}", (t2 - t1).as_secs_f64());

    println!("{}", h2.len());
}
```

Is this expected behavior for an `FxHashMap`? Is there a way to fix it without going through a custom (and space-inefficieent) serialization/deserialization?

Note: I have reported the essentially same issue is reported to the `fxhash` crate, which also provides an `FxHashMap` with the same deserialization behavior.

### LINKED ISSUE #15: Provide public API for initial seed of `FxHasher`
`FxHasher` can currently only be created with the 0 value for `hash`. It would be useful to have a constructor that constructs it with the explicitly provided seed.

As described in #14, certain operations on `FxHashMap` could be made faster by making the map unstable, i.e. changing the seed of the hash function for each new map. However,  `rustc_hash::FxHasher` doesn't offer a way to create the hasher with the specified initial hash value. As a workaround, we can use `write_u32()` to modify the value:

```rust
// BuildHasherRandom is created with random state, like RandomState in the
// stlib - https://bit.ly/39pEvnN
#[derive(Debug, Clone)]
struct BuildHasherRandom(u32);

impl BuildHasher for BuildHasherRandom {
    type Hasher = FxHasher;

    fn build_hasher(&self) -> FxHasher {
        // workaround: create `FxHasher` with the random seed from self.0
        let mut hasher = FxHasher::default();
        hasher.write_u32(self.0);
        hasher
    }
}
```

It's not obvious that the compiler is able to optimize `write_u32(n)` with a known starting value of `hasher.hash`. Ideally `build_hasher()` should be able to explicitly provide the seed to the `FxHasher` constructor, such as:

```rust
    fn build_hasher(&self) -> FxHasher {
        FxHasher::from_seed(self.0)
    }
```

Note that we need `from_seed` only to implement randomization to fix #14. If an equivalent of `RandomState` were built into `rustc_hash`, we would no longer need the `from_seed` functionality.

### PR DESCRIPTION
**Pull request by WaffleLapkin**
tl;dr: this PR adds the following APIs:
```rust
impl FxHasher {
    pub fn with_seed(seed: usize) -> FxHasher;
}

// Everything below is under cfg(feature = "rand")

pub struct FxRandomState { ... }

impl FxRandomState {
    pub fn new() -> FxRandomState;
}

impl BuildHasher for FxRandomState { ... }
```

This adds ways to create `FxHasher` with a custom or random seed.

Resolves #14
Resolves #15

Helps with #17, though different default seed may still be a good idea.

### CODE REVIEW

**Comment by PureWhiteWu** [comment]
Hi, is there any plan for this pr to be merged?
I think this is quite important.

File: src/lib.rs
Comment (Nilstrieb) [inline src/lib.rs:54]:
```suggestion
pub type FxHashMapRand<K, V> = HashMap<K, V, FxRandomState>;
```
I'd prefer writing out the `a`, or maybe dropping these aliases alltogether.

Comment (WaffleLapkin) [reply src/lib.rs]: 🆗

**Review by Nilstrieb — CHANGES_REQUESTED** [review]
Can you add some tests for this?
One useful test would be to create two `FxRandomState` and ensure their seeds aren't the same.
Another possible test would be creating two `FxRandomState` on two different threads and ensuring their seeds aren't the same (this is not *guaranteed* to always pass, but should in practice always pass).
Another useful test would be for `FxHasher::with_seed` to use it with two different seeds (and the default) and ensure that hashing some integer with it produces different results in all cases.

**Comment by WaffleLapkin** [comment]
@Nilstrieb are the tests I added match what you asked for?

I also added a `BuildHasher` impl you can provide a seed for, since there wasn't a way to do this without making your own type... Does it look reasonable to you?

File: src/lib.rs
Comment (Nilstrieb) [inline src/lib.rs:35]: Can we add a newline between the modules to make it more obvious that only the former is cfged?

**Review by Nilstrieb — APPROVED** [review]
r=me after style nit (also on the re-exports)

**Comment by WaffleLapkin** [comment]
Uuuuhhhhh is the master branch not protected? D:

**Comment by Nilstrieb** [comment]
I think it is, but I approved your PR, and GitHub does not dismiss approvals after pushing

**Comment by WaffleLapkin** [comment]
@Nilstrieb I merged the PR before CI passed (or at least it looked like I did, maybe it happened at the same time?). also fyi github has an option to dismiss approvals after pushing

**Comment by Nilstrieb** [comment]
interesting.. i wouldn't wanna dismiss approvals, that's just annoying, but if it's not gated on CI that would be a bit bad. not that it matters too much

**Pull request merged**
